### PR TITLE
Dont stop mining loop in case race condition lead to using output that was just spent

### PR DESCRIPTION
**Pull request by noescape00**
It seems that due to the race condition we might try to use output which was just spent on consensus chain

https://dev.azure.com/Stratisplatformuk/StratisBitcoinFullNode/_workitems/edit/4172

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs [hunk 1]
```diff
@@ -165,15 +165,18 @@ private async Task CreateBlocksAsync()
...
                     this.logger.LogInformation("Miner failed to mine block due to: '{0}'.", ce.ConsensusError.Message);
                 }
                 catch (Exception exception)
                 {
                     this.logger.LogCritical("Exception occurred during mining: {0}", exception.ToString());
-                    break;
+
+                    // This can happen due to the race condition, we shouldn't break mining loop because of this.
+                    // TODO: Find a better way to do this.
+                    if (exception.Message != "Output has already been spent.")
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs]: Can't you catch a better exception? what is throwing this?

Comment (noescape00) [reply src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs]: That comes deep from the smart contracts engine so I'd not dare to refactor it to push a better exception.

**Comment by codingupastorm** [comment]
@dangershony this isn't only an SC error, occurs in PoS too if you spend a utxo in a block that you just receive and validate, whilst in the process of trying to stake with the same utxo. The BlockDefinitions rely on `ICoinView` and there are race conditions as it's accessed by both consensus and miner.

**Comment by noescape00** [comment]
```
It seems the ConsensusManager makes the assumption that the miner will always catch consensus exceptions so I think the PoAMiner should just go with that for now:
```

It does catch the exception. The question is: should we stop mining loop in case we have that particular exception (missing output). And it looks like we shouldn't because it can happen due to the race condition.

**Comment by codingupastorm** [comment]
Right, but #3842 already caters for this error and others. There is another example in the PR Review column for a different error that also needs to be caught, we should use the same functionality as PoS and catch all consensus exceptions

**Comment by noescape00** [comment]
>Right, but #3842 already caters for this error and others. 

Commented on it. 

Looks like we have 2 PRs that address roughly the same problem. I don't mind if ether goes in but we need to agree on how we handle those errors. 

My position is that we shouldn't ignore all consensus errors, only those that we know can occur normally and not as a consequence of a bug.

**Pull request closed without merging**